drivers/mbox/mbox_ti_omap.c

```diff
@@ -0,0 +1,196 @@
+
+	for (int i_channel = 0; i_channel < MAILBOX_MAX_CHANNELS; i_channel++) {
+		if (!data->channel_enable[i_channel]) {
+			continue;
+		}
+
+		if ((flags & MAILBOX_IRQ_NEWMSG(i_channel))) {
+			data->received_data = regs->message[i_channel];
+			struct mbox_msg msg = {(const void *)&data->received_data,
+					       MAILBOX_MBOX_SIZE};
+
+			if (data->cb[i_channel]) {
+				data->cb[i_channel](dev, i_channel, data->user_data[i_channel],
+						    &msg);
+				irq_regs->status_clear = MAILBOX_IRQ_NEWMSG(i_channel);
+			}
+
+		}
+	}
+}
+
+static int omap_mailbox_send(const struct device *dev, uint32_t channel, const struct mbox_msg *msg)
+{
+	uint32_t __aligned(4) data32;
+	struct omap_mailbox_data *data = dev->data;
+	volatile struct omap_mailbox_regs *regs = (struct omap_mailbox_regs *)data->mapped_base;
+
+	if (channel >= MAILBOX_MAX_CHANNELS) {
+		return -EINVAL;
+	}
+
+	if (!msg) {
+		if (!regs->fifo_status[channel]) {
+			regs->message[channel] = 0;
+		}
+		return 0;
+	}
+
+	if (msg->size != MAILBOX_MBOX_SIZE) {
+		return -EMSGSIZE;
+	}
+
+	memcpy(&data32, msg->data, msg->size);
+
+	if (!regs->fifo_status[channel]) {
+		regs->message[channel] = data32;
+	}
+
+	return 0;
+}
+
+static int omap_mailbox_register_callback(const struct device *dev, uint32_t channel,
+					  mbox_callback_t cb, void *user_data)
+{
+	struct omap_mailbox_data *data = dev->data;
+
+	if (channel >= MAILBOX_MAX_CHANNELS) {
+		return -EINVAL;
+	}
+
+	data->cb[channel] = cb;
+	data->user_data[channel] = user_data;
+
+	return 0;
+}
+
+static int omap_mailbox_mtu_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return MAILBOX_MBOX_SIZE;
+}
+
+static uint32_t omap_mailbox_max_channels_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return MAILBOX_MAX_CHANNELS;
+}
+
+static int omap_mailbox_set_enabled(const struct device *dev, uint32_t channel, bool enable)
+{
+	struct omap_mailbox_data *data = dev->data;
+	const struct omap_mailbox_config *cfg = dev->config;
+	volatile struct omap_mailbox_regs *regs;
+	uint32_t irqstatus;
+
+	regs = (struct omap_mailbox_regs *)data->mapped_base;
+	if (channel >= MAILBOX_MAX_CHANNELS) {
+		return -EINVAL;
+	}
+
+	irqstatus = regs->irq_regs[cfg->usr_id].enable_set;
+	irqstatus |= MAILBOX_IRQ_NEWMSG(channel);
+	regs->irq_regs[cfg->usr_id].enable_set = irqstatus;
+	data->channel_enable[channel] = enable;
+
+	return 0;
+}
+
+static const struct mbox_driver_api omap_mailbox_driver_api = {
+	.send = omap_mailbox_send,
+	.register_callback = omap_mailbox_register_callback,
+	.mtu_get = omap_mailbox_mtu_get,
+	.max_channels_get = omap_mailbox_max_channels_get,
+	.set_enabled = omap_mailbox_set_enabled,
+};
+
+#define MAILBOX_INSTANCE_DEFINE(idx)	\
+	static struct omap_mailbox_data omap_mailbox_##idx##_data;	\
+	const static struct omap_mailbox_config omap_mailbox_##idx##_config = {	\
+		.usr_id =  DT_INST_PROP(idx, usr_id),	\
+	};	\
+	static int omap_mailbox_##idx##_init(const struct device *dev)	\
```

Comment:
you could save the idx number in the `omap_mailbox_##idx##_config` and call a generic init function. Saves some code size when multiple mailbox drivers are enabled... your approach works too.

Reply:
Will fix this, this is much better as few other devices have multiple (10+) mailbox instances.

Reply:
I have skipped this for now, if it is okay I will fix it later when adding multiple instances